src/lender/coordinator.sol

```diff
@@ -125,97 +142,264 @@ contract EpochCoordinator is Ticker, Auth  {
+        // we might accept it as an improvement
+        if (gotValidPoolConSubmission == false) {
+            return _improvementScoreCase(seniorRedeem, juniorRedeem, juniorSupply, seniorSupply);
+        }
 
-        bestSubScore = score;
+        // proposed solution doesn't satisfy the pool constraints but a previous submission did
+        // previous solutions satisfied all constraints
+        return -3;
+    }
+
+    function abs(uint x, uint y) public view returns(uint delta) {
+        if(x == y) {
+            // todo add explanation
+            return 1;
+        }
+
+        if(x > y) {
+            return safeSub(x, y);
+        }
+        return safeSub(y, x);
+    }
+
+    function checkRatioInRange(Fixed27 memory ratio, Fixed27 memory minRatio,
+        Fixed27 memory maxRatio) public view returns (bool) {
+        if (ratio.value >= minRatio.value && ratio.value <= maxRatio.value ) {
+            return true;
+        }
+        return false;
+    }
+
+    function _improvementScoreCase(uint seniorRedeem, uint juniorRedeem,
+        uint juniorSupply, uint seniorSupply) internal returns(int) {
+        Fixed27 memory currSeniorRatio = Fixed27(calcSeniorRatio(safeAdd(epochSeniorBalance, epochSeniorDebt),
+            epochNAV, epochReserve));
+
+        if(bestSubScore == 0) {
+            // define no orders score as benchmark if no previous submission exists
+            bestSubScore = scoreImprovement(currSeniorRatio, currSeniorRatio, epochReserve);
+        }
+
+        uint newReserve = calcNewReserve(seniorRedeem, juniorRedeem, seniorSupply, juniorSupply);
 
-        // solution is new best => 0
+        Fixed27 memory newSeniorRatio = Fixed27(calcSeniorRatio(calcNewSeniorAssetValue( seniorRedeem, seniorSupply,
+            epochSeniorDebt, epochSeniorBalance, newReserve), epochNAV, newReserve));
+
+        uint score =  scoreImprovement(newSeniorRatio, currSeniorRatio, newReserve);
+
+        if (score < bestSubScore) {
+            // solution is not the best => -1
+            return -1;
+        }
+
+        // solution doesn't satisfy all pool constraints but improves the current violation
+        saveNewOptimum(seniorRedeem, juniorRedeem, juniorSupply, seniorSupply, score);
         return 0;
     }
 
+    // returns the normalized distance (maxReserve = ONE) from the newReserve and maxReserve/2
+    function scoreDistanceReserve(Fixed27 memory newReserve_) public view returns (uint score) {
+        return rmul(1000, rdiv(ONE, abs(safeDiv(ONE, 2), newReserve_.value)));
+    }
+
+    function scoreImprovement(Fixed27 memory newSeniorRatio, Fixed27 memory currSeniorRatio,
+        uint newReserve_) public  returns(uint) {
+        (Fixed27 memory minSeniorRatio, Fixed27 memory maxSeniorRatio) = assessor.seniorRatioBounds();
+
+        // normalize reserve by defining maxReserve as ONE
+        Fixed27 memory normalizedNewReserve = Fixed27(rdiv(newReserve_, assessor.maxReserve()));
```

Comment:
Maybe  this value can be computed in the normalizedNewReserve function, as it is not used in this function

Reply:
If we calculate it here we only need to calculate it once.

If we calculate it in the method itself, the calculation would be performed twice